Roadmap page

laymonage opened the pull request:
## How to run

- Checkout the branch
- Might need to run `make setup` to ensure your local build (if any) is up to date with the latest dependencies
- Follow the instructions on the README
  - `make start`, `make runserver`, `make migrate`, `make pull-production-data`, `make pull-production-media`, `make superuser`
- Create a new GitHub personal access token
  - ~~https://github.com/settings/tokens/new~~ https://github.com/settings/personal-access-tokens/new
  - No access is necessary at all, leave all options unchecked
  - Note that the "Fine-grained personal access tokens" does not work with the GraphQL API ([see docs](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token#personal-access-tokens-classic)), so use "Personal access tokens (classic)" instead
    - **EDIT:** This just in! https://github.blog/changelog/2023-04-27-graphql-improvements-for-fine-grained-pats-and-github-apps/
- Edit `settings/local.py` and set `GITHUB_ACCESS_TOKEN = "ghp_foobar"`
- Open the website's Wagtail admin, go to Settings > GitHub Roadmap > Import all milestones
- Open the v5.1 milestone
- Uncheck the Publish field on the "Provisional version number" item
- Add [this link](https://www.figma.com/proto/avNL91NlYIbxjvfKyO8pvp/Wagtail-Marketing?node-id=164-1037&viewport=1021%2C665%2C1&scaling=scale-down-width&starting-point-node-id=164%3A1037&show-proto-sidebar=1&hide-ui=1&) to the Treeless item
- Go to Settings > Redirects and delete/update the `/roadmap` redirect
- Create a new Roadmap page under the Home page
- Fill out the form
  - Sponsorship page: choose the "Sponsor Wagtail" page
  - Fine print:
    ```
    Future version numbers are provisional - these may receive a major version number increment, depending on the extent of backwards-incompatible changes and significant UI differences.
    ```
  - Body: add a standalone CTA block
    - CTA text: `Sponsor Wagtail`
    - CTA page: choose the "Sponsor Wagtail" page
    - Short description:
      ```
      Help build Wagtail enhancements and support the long-term health of the project
      ```

## Notes

We should think about how the access token should be created for production.

The most proper way to do this would be to create a GitHub App under the Wagtail organisation, and then authenticate as the app. However, this is much more complicated as we'll need to:
- [Create a GitHub App](https://docs.github.com/en/apps/creating-github-apps/setting-up-a-github-app/creating-a-github-app)
- [Generate a JWT](https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/generating-a-json-web-token-jwt-for-a-github-app)
- Use the JWT to [generate an access token as a GitHub App **installation**](https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/authenticating-as-a-github-app-installation).
- Use the token in our GraphQL requests.
  - This access token expires in 1 hour, so this would need to be generated on-the-fly. This means that instead of storing the token in the environment variable, we would need to store the App ID and the App's private key instead, so that we can generate the JWT at runtime.

Alternatively, the much easier way would be to use someone's account (or an empty account) to generate a personal access token without any access, which I've done here.

**EDIT: Now that [fine-grained access tokens can be used on the GraphQL API](https://github.blog/changelog/2023-04-27-graphql-improvements-for-fine-grained-pats-and-github-apps/), we can create an access token with the Wagtail organisation as the owner, and with read-only access to public repositories (without any permissions configured).**

thibaudcolas commented on wagtailio/roadmap/models.py:
```diff
@@ -0,0 +1,185 @@
+
+
+class Milestone(ClusterableModel):
+    publish = models.BooleanField(default=True)
```
I don’t understand what this `publish` is for either – in what scenario would we be using this to override the GitHub data? If there’s one then I think it’d be worth a comment, or have the help text explain the scenario.

thibaudcolas replied:
I don’t understand what this `publish` is for either – in what scenario would we be using this to override the GitHub data?

thibaudcolas reviewed (changes requested):
All working pretty well @laymonage. There are quite a few changes I’ve suggested here though none too big.

For the API token, I think my preference would be for you to add a field to the import form, so people who use this can create and submit their own token at the point it’s needed.

It’s a bit more hassle but means we won’t need to manage a new GitHub account just for this, which feels like a lot of admin for little value.

laymonage commented on wagtailio/roadmap/templates/roadmap/roadmap_page.html:
```diff
@@ -0,0 +1,83 @@
+                                        </span>
+                                        <svg class="icon" aria-hidden="true" viewBox="0 0 100 100">
+                                            <use xlink:href="#arrow-up-right-from-square">
+                                            </svg>
```
It was already formatted with `djhtml`, but I missed the closing `</use>` tag. We already have an `icon.html` for the double (fg-bg icons), so I hesitated adding yet another icon.

I've changed the component template to accept a `variant` variable that uses this single SVG markup `if variant == "simple"`.

laymonage commented on wagtailio/roadmap/templates/roadmap/roadmap_page.html:
```diff
@@ -0,0 +1,83 @@
+                                        {% endfor %}
+                                    </div>
+                                </div>
+                                {% if item.needs_sponsorship or item.sponsorship_url %}
```
No, we don't want to use `and` because we want to link to the sponsorship page if the item has no `sponsorship_url`.

I used `or` in case some items have a custom `sponsorship_url` but no `needs sponsorship` label on GitHub. But you're right, let's use GitHub as the source of truth. This means that the custom `sponsorship_url` takes no effect unless the item has `needs sponsorship` label on GitHub.

thibaudcolas replied:
My bad, indeed what I’d have expected is to check `needs_sponsorship` only.

laymonage commented on wagtailio/roadmap/templates/roadmap/roadmap_page.html:
```diff
@@ -0,0 +1,83 @@
+                                    </div>
+                                </div>
+                                {% if item.needs_sponsorship or item.sponsorship_url %}
+                                    {% pageurl page.sponsorship_page item.url as sponsor_url %}
```
It's a fallback for the `pageurl` tag if the `sponsorship_page` foreign key is unset, so that it doesn't crash when it tries to compute from `None`. This `item.url` is only used if both the `sponsorship_page` FK on the page is unset and the item has no `sponsorship_url`. It's just for safeguarding more than anything.

laymonage commented on wagtailio/static/sass/abstracts/_variables.scss:
```diff
@@ -25,10 +25,16 @@ $font--code: 'courier new', monospace;
 // Font weights
 $weight--bold: 700;
+$weight--semibold: 600;
+$weight--medium: 500;
```
I've changed this to only add `semibold` and without the variable font.

laymonage commented:
@thibaudcolas I've made adjustments per your review:

- Removed the `publish` fields. We don't need this after all, as the solution to the "Provisional version number" thing is easy: just don't add it to the Milestone. It can still be added to the Project if we want to.
- Added the `intro` field and use the `hero.html` template
- Added the `semibold` weight only, without the variable font file.
- Made the whole milestone item heading area clickable.
- Added the option to input a GitHub access token during import.
- Fixed accessibility issues and made other refactoring/improvements as mentioned.

thibaudcolas reviewed (approved):
All looking great! Only thing I’d have suggested doing differently is to combine the migrations so it looks neater, but doesn’t really matter for a project of this size.

laymonage commented:
> All looking great! Only thing I’d have suggested doing differently is to combine the migrations so it looks neater, but doesn’t really matter for a project of this size.

Yes, I considered doing that but I thought it would make it harder for you to do another review as you’d need to make sure it could be run locally after the initial migration 😅
